patches/tModLoader/Terraria/Social/Base/SocialBrowserModule.cs

```diff
@@ -0,0 +1,206 @@
+	{
+	}
+}
+
+public interface SocialBrowserModule
+{
+	public bool Initialize();
+
+	/////// Management of Browser Items ///////////////////////////////////////////
+
+#pragma warning disable CS8424 // I know [EnumeratorCancellation] has no effect, but it's placed here to remember to add it to async implementations
+	public IAsyncEnumerable<ModDownloadItem> QueryBrowser(QueryParameters queryParams, [EnumeratorCancellation] CancellationToken token = default);
+#pragma warning restore CS8424
+
+	public ModDownloadItem[] DirectQueryItems(QueryParameters queryParams);
+
+	/////// Display of Browser Items ///////////////////////////////////////////
+
+	public string GetModWebPage(ModPubId_t item);
+
+	/////// Management of Local Install ///////////////////////////////////////////
+	public bool GetModIdFromLocalFiles(TmodFile modFile, out ModPubId_t item);
+
+	// Needed for ensuring that the 'Update All' button works correctly. Without caching the mod browser locks out on the update all button
+	internal ModDownloadItem[] CachedInstalledModDownloadItems { get; set; }
+
+	public ModDownloadItem[] DirectQueryInstalledMDItems(QueryParameters qParams = new QueryParameters()) {
+		var mods = GetInstalledMods();
+		var listIds = new List<ModPubId_t>();
+
+		foreach (var mod in mods) {
+			if (GetModIdFromLocalFiles(mod.modFile, out var id))
+				listIds.Add(id);
+		}
+
+		qParams.searchModIds = listIds.ToArray();
+
+		return DirectQueryItems(qParams);
+	}
+
+	public ModDownloadItem[] GetInstalledModDownloadItems()
+	{
+		if (CachedInstalledModDownloadItems == null) {
+			CachedInstalledModDownloadItems = DirectQueryInstalledMDItems();
+		}
+
+		return CachedInstalledModDownloadItems;
+	}
+
+	/////// Specialty Internal LocalMod related Methods ///////////////////////////////////////////
+	public bool DoesAppNeedRestartToReinstallItem(ModPubId_t modId);
+
+	internal bool DoesItemNeedUpdate(ModPubId_t modId, LocalMod installed, Version webVersion);
+
+	internal IReadOnlyList<LocalMod> GetInstalledMods();
+
+	internal LocalMod IsItemInstalled(string slug)
+	{
+		return GetInstalledMods().Where(t => t.Name == slug).FirstOrDefault();
+	}
+
+	/////// Management of Downloads ///////////////////////////////////////////
+
+	/// <summary>
+	/// Downloads all UIModDownloadItems provided.
+	/// </summary>
+	internal Task SetupDownload(List<ModDownloadItem> items, int previousMenuId)
+	{
+		// Can't update enabled items due to in-use file access constraints
+		// Mod unloading needs to occur off the main thread o avoid UI impacts/soft-freeze
+		var needFreeInUseMods = items.Any(item => { item.UpdateInstallState(); return item.IsEnabled; });
+		if (needFreeInUseMods) {
+			Interface.loadMods.suppressAutoLoad = true;
+			Main.menuMode = Interface.loadModsID;
+			var unload = Task.Run(() => ModLoader.ModLoader.Unload());
+			return unload.ContinueWith(delegate { DownloadRunner(items, previousMenuId, needFreeInUseMods); });
```

Comment:
This is actually quite dangerous. `Unload` doesn't leave tML in a valid state on its own if it returns `false`. 

I'd prefer if you did a reload which temporarily disables mods, and ran the download runner in `OnSuccessfulLoad`